### PR TITLE
fix: refactor `ibc-derive` to handle context with generic types and projects dependent on `ibc-core`

### LINKED ISSUE #910: `ClientState` derivation does not work with contexts generic over types and/or lifetimes
## Summary
From @yito88

The `ClientState` derivation isn't working well with contexts generic over types/lifetimes. 

## Details
The following case, for instance, does not work:

```rust
#[derive(ClientState)]
#[generics(
    ClientValidationContext = IbcActions::<'_, C>,
    ClientExecutionContext = IbcActions::<'_, C>,
)]
enum AnyClientState {
    Tendermint(TmClientState),
}
```

As well, this has been experienced in the context of Sovereign SDK integration (see [here](https://github.com/informalsystems/sovereign-ibc/blob/fbd4db5615ea931dbc065e43dace4da99b4df081/modules/sov-ibc/src/clients/context.rs#L22-L25))

## Version
v0.44.1

### LINKED ISSUE #999: `ibc-derive` does not work if only depend on `ibc-core`
## Description
The current implementation of `ibc-derive` exclusively supports projects dependent on the `ibc` meta-crate, where [the required types](https://github.com/cosmos/ibc-rs/blob/4421b0dcd6eab6c914b8c5a7d45817b1edb6b2ce/ibc-derive/src/utils.rs#L11-L75) for implementing `ClientState` and `ConsensusState` traits are imported directly from `ibc`.
This poses a limitation as not all projects necessarily rely on `ibc`. To enhance the utility of `ibc-derive`, it should support projects that may depend on the `ibc-core` (or even `ibc-core-client`) crates.

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #910
Closes: #999 
______

### PR author checklist:
- [X] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [X] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-derive/src/utils.rs [hunk 1]
```diff
@@ -4,73 +4,109 @@ use syn::{Path, Variant};
...
+    pub fn new_ibc_core() -> Self {
+        Self {
+            prefix: quote! {::ibc_core},
+        }
+    }
+
+    pub fn prefix(&self) -> &TokenStream {
+        &self.prefix
+    }
+}
```
Comment (rnbguy) [inline ibc-derive/src/utils.rs]: IMO this seems like `ibc-derive` requires `ibc_core`.

Can't we add that in `ibc_derive` dependency and exports and use it here as `ibc_derive::ibc_core` or something like that?

File: ibc-derive/src/client_state.rs [hunk 2]
```diff
@@ -1,29 +1,69 @@
...
 use traits::client_state_validation::impl_ClientStateValidation;
 
-#[derive(FromDeriveInput)]
-#[darling(attributes(generics))]
+use crate::utils::Imports;
+
+const MISSING_ATTR: &str = "must be annotated with #[validation(<your ClientValidationContext>) and #[execution(<your ClientExecutionContext>)]";
+const MISSING_VALIDATION_ATTR: &str = "missing #[validation(<your ClientValidationContext>)]";
+const MISSING_EXECUTION_ATTR: &str = "missing #[execution(<your ClientExecutionContext>)]";
+const INVALID_ATTR: &str = "invalid attribute annotation!";
```
Comment (rnbguy) [inline ibc-derive/src/client_state.rs]: can we remove the exclamation?

File: ibc-derive/src/client_state/traits/client_state_execution.rs [hunk 3]
```diff
@@ -11,45 +11,62 @@ pub(crate) fn impl_ClientStateExecution(
...
+        Some(segment) => match segment.arguments {
+            syn::PathArguments::AngleBracketed(ref gen) => {
+                let Gen = gen.args.clone();
+
+                quote! { impl<#Gen> }
+            }
+            _ => quote! { impl },
+        },
+        None => panic!("Invalid ClientExecutionContext type"),
+    };
```
Comment (rnbguy) [inline ibc-derive/src/client_state/traits/client_state_execution.rs]: does it make sense to make this a method? as we reuse the same logic for validation too.

File: ibc-derive/src/lib.rs [hunk 4]
```diff
@@ -16,21 +16,48 @@ use client_state::client_state_derive_impl;
...
 
-    let output = client_state_derive_impl(ast);
+    let imports = Imports::new_ibc();
+
+    let output = client_state_derive_impl(ast, &imports);
+
+    RawTokenStream::from(output)
+}
+
+#[proc_macro_derive(IbcCoreClientState, attributes(validation, execution))]
```
Comment (rnbguy) [inline ibc-derive/src/lib.rs]: I have added [a comment](https://github.com/cosmos/ibc-rs/pull/1037#discussion_r1447239845) about the possibility of not maintaining two duplicate derive macros here for client state and later for consensus state.

**Review by rnbguy — CHANGES_REQUESTED** [review]
really awesome PR @Farhad-Shabani ! Really like the solution. :+1:

I asked for some clarifications in the comments.

**Comment by rnbguy** [comment]
If I have a context with trait-bounded generics as following:

```rs
struct MyIbcContext<T: MyTrait> { ... }
```

How do I the trait-bound its generics while deriving `IbcClientState`? The following doesn't work.

```rs
#[derive(IbcClientState)]
#[validation(MyIbcContext<T: MyTrait>)]
#[execution(MyIbcContext<T: MyTrait>)]
enum AnyClientState { ... }
```

File: ibc-derive/src/utils.rs (hunk 1 above)
Comment (Farhad-Shabani) [reply ibc-derive/src/utils.rs]: Until expansion takes place, the inclusion of `ibc_core` isn't happening. However, I've adjusted things by relocating each set of derive macros to their relevant meta-crate. This ensures that when working with `::ibc`, we won't unintentionally import `::ibc_core`, and vice versa. See if you like it?

File: ibc-derive/src/lib.rs (hunk 4 above)
Comment (Farhad-Shabani) [reply ibc-derive/src/lib.rs]: https://github.com/cosmos/ibc-rs/pull/1037#discussion_r1447532259

**Comment by Farhad-Shabani** [comment]
> How do I the trait-bound its generics while deriving `IbcClientState`? The following doesn't work.


What about this?

```rust
#[derive(IbcClientState)]
 #[validation(MyIbcContext<T>)]
 #[execution(MyIbcContext<T>)]
 enum AnyClientState { ... }
```

**Comment by rnbguy** [comment]
The trait bounds are needed when we are deriving the `ClientState{Validation, Execution}` :sweat_smile:

**Comment by Farhad-Shabani** [comment]
> The trait bounds are needed when we are deriving the `ClientState{Validation, Execution}` 😅

Ahh, that's right. That's a new case!

File: ibc-derive/src/utils.rs (hunk 1 above)
Comment (rnbguy) [reply ibc-derive/src/utils.rs]: Yea, this solution is much cleaner ! Thanks :pray:

**Review by rnbguy — APPROVED** [review]
Thanks @Farhad-Shabani ! Looks awesome. Great work :tada:

**Pull request merged**